src/main/scala/mesosphere/marathon/raml/PodStatusConversion.scala

```diff
@@ -0,0 +1,116 @@
+
+import mesosphere.marathon.core.instance.{ Instance, InstanceStatus }
+import mesosphere.marathon.core.pod.PodDefinition
+import mesosphere.marathon.state.RunSpec
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Seq
+
+trait PodStatusConversion {
+
+  import PodStatusConversion._
+
+  /**
+    * generate a pod instance status RAML for some instance.
+    */
+  @throws[IllegalArgumentException]("if you provide a non-pod `spec`")
+  implicit val podInstanceStatusRamlWriter: Writes[Source, PodInstanceStatus] = Writes[Source, PodInstanceStatus] { src =>
+
+    val (spec, instance) = src
+
+    // BLOCKED: need capability to get the container name from Task somehow, current thinking is that Task.Id will
+    // provide such an API.
+
+    val pod: PodDefinition = spec match {
+      case x: PodDefinition => x
+      case _ => throw new IllegalArgumentException(s"expected a pod spec instead of $spec")
+    }
+
+    assume(
+      pod.id == instance.instanceId.runSpecId,
+      s"pod id ${pod.id} should match spec id of the instance ${instance.instanceId.runSpecId}")
+
+    // TODO(jdef) associate task w/ container by name, allocated host ports should be in relative order
+    // def endpoints = instance.tasks.map(_.launched.map(_.hostPorts))
+
+    def containerStatus: Seq[ContainerStatus] = instance.tasks.map { task =>
+      val since = task.status.startedAt.getOrElse(task.status.stagedAt).toOffsetDateTime // TODO(jdef) inaccurate
+
+      // some other layer should provide termination history
+      // TODO(jdef) message, conditions
+      ContainerStatus(
+        name = task.taskId.mesosTaskId.getValue, //TODO(jdef) pods this is wrong, should be the container name from spec
+        status = task.status.taskStatus.toMesosStateName,
+        statusSince = since,
+        containerId = task.launchedMesosId.map(_.getValue),
+        endpoints = Seq.empty, //TODO(jdef) pods, report endpoint health, allocated host ports here
+        lastUpdated = since, // TODO(jdef) pods fixme
+        lastChanged = since // TODO(jdef) pods.fixme
+      )
+    }(collection.breakOut)
+
+    val derivedStatus: PodInstanceState = instance.state.status match {
+      case InstanceStatus.Created | InstanceStatus.Reserved => PodInstanceState.Pending
+      case InstanceStatus.Staging | InstanceStatus.Starting => PodInstanceState.Staging
+      case InstanceStatus.Error | InstanceStatus.Failed | InstanceStatus.Finished | InstanceStatus.Killed |
+        InstanceStatus.Gone | InstanceStatus.Dropped | InstanceStatus.Unknown | InstanceStatus.Killing |
+        InstanceStatus.Unreachable => PodInstanceState.Terminal
+      case InstanceStatus.Running =>
+        if (instance.state.healthy.getOrElse(true)) PodInstanceState.Stable else PodInstanceState.Degraded
+    }
+
+    val networkStatus: Seq[NetworkStatus] = instance.tasks.flatMap { task =>
+      task.mesosStatus.filter(_.hasContainerStatus).fold(Seq.empty[NetworkStatus]) { mesosStatus =>
```

Comment:
withFilter is lazy, filter computes the collection immediately. If you're doing a chain of transforms, in theory, views are better. I have a change incoming where you can actually flip over to the java8 stream api transparently - and in there you explicitly have "terminal operations" which "complete the stream". scala views are supposed to work similarly to this.